test/type/protocol_types.swift

```diff
@@ -138,3 +138,8 @@ struct BadSubscript {
     set {}
   }
 }
+
+struct OuterGeneric<T> {
+  func contextuallyGenericMethod() where T == HasAssoc {}
+  // expected-error@-1 {{protocol 'HasAssoc' can only be used as a generic constraint because it has Self or associated type requirements}}
```

Comment:
Since `A == B` is also a constraint, perhaps we should clarify the error message to say "generic conformance constraint" or simply "conformance constraint"